File: src/main/java/com/coding/programmers/level2/타겟넘버.java
package com.coding.programmers.level2;

public class 타겟넘버 {
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.solution(new int[]{1, 1, 1, 1, 1}, 3));
    }

    static class Solution {
        public int solution(int[] numbers, int target) {
            return dfs(0, 0, numbers, target);
        }

        int dfs(int sum, int idx, int[] nums, int target) {
            if (idx == nums.length && sum == target) {
                return 1;
            }
            if (idx >= nums.length) {
                return 0;
            }
            return dfs(sum + nums[idx], idx + 1, nums, target) + dfs(sum - nums[idx], idx + 1, nums, target);
        }
    }
}

File: src/main/kotlin/com/coding/programmers/graph/dfs/타겟넘버.kt
package com.coding.programmers.graph.dfs
fun main() {
    val r = 타겟넘버().solution(intArrayOf(
        1,1,1,1,1
    ), 3)

    println(r)
}
class 타겟넘버 {
    fun solution2(numbers: IntArray, target: Int): Int {
        return numbers.fold(listOf(0)) {
            list, i ->
            list.run {
                map { it + i } + map { it - i }
            }
        }.count { it == target }
    }

    fun solution(numbers: IntArray, target: Int): Int {
        return dfs(0, 0, numbers, target)
    }
    fun dfs(sum: Int, idx: Int, nums:IntArray, target: Int) : Int {
        if (idx == nums.size && sum == target) {
            return 1
        }
        if (idx >= nums.size) {
            return 0
        }
        return dfs(sum + nums[idx], idx + 1, nums, target) + dfs(sum - nums[idx], idx + 1, nums, target)
    }
}